### PR TITLE
Add a `set_seed` function

### LINKED ISSUE #56: Add the possibility to seed the generations
Generative models return samples from a distribution; if no seed is set they should return different results each time they are called. The current caching implementation however does not take the random state into account. It is only based on the other inputs, so two successive runs will return the same value even though no seed has been set. This behavior is unintuitive and should be corrected.

Instead, we should add a mechanism so users can set the seed globally. If the seed is set the cache will be hit at the next run, otherwise a new result will be returned. This is however made slightly complicated by the fact that it is not possible to pass a seed value to some API providers like OpenAI and Anthropic, so they will generate a different value each time they are called when the temperature is different from 0.

I have thought of a possible workaround. We can manually define cache keys in the form:

```
random_seed + model_name + call # + argument values
```

- The `random_seed` can be set globally by the user with `outlines.seed(42)`, otherwise it is drawn at random when requested. This ensures that consecutive runs will not hit the cache when no seed is set.
- `model_name` is here to ensure that we're not hitting the cache if we've changed the model used for generation between two runs.
- `argument_values` so the cache is not hit when the same model is called twice with different arguments.
- `call #` so we can cache calls to closed APIs with non zero temperature and  repeated calls within the same program execution to yield different results.

### PR DESCRIPTION
**Pull request by brandonwillard**
This PR adds a basic `set_seed` function.  It doesn't currently handle the caching mentioned in #56 (more on that later).

- [ ] Test other API providers

Closes #56

### CODE REVIEW

File: outlines/__init__.py [hunk 1]
```diff
@@ -1,10 +1,38 @@
...
+        import torch
+
+        torch.manual_seed(seed)
+        torch.cuda.manual_seed_all(seed)
+        # ^^ safe to call this function even if cuda is not available
+
+    with suppress(ImportError):
+        import tensorflow
+
+        tensorflow.random.set_seed(seed)
```
Comment (brandonwillard) [inline outlines/__init__.py:38]: Importing these just in case a user uses a provider backed by them seems awfully wasteful.  We could delay seeding until we know exactly which provider and backend is being used, but that sounds like it could get complicated.

File: outlines/caching.py [hunk 2]
```diff
@@ -56,3 +56,11 @@ def clear_cache():
...
     global memory
     memory.storage.clear()
+
+
+def enable_cache():
+    """Enable the cache."""
+    global memory
+
+    if isinstance(memory, NoCache):
+        memory = Cache(serializer=JSONSerializer(), storage=LocalFileStorage(cache_dir))
```
Comment (brandonwillard) [inline outlines/caching.py:66]: I left this in from when I was attempting to test the seeding with and without caching in `tests/models/test_hf_transformers.py`; however, it looks like the caching approach might be restricting this sort of dynamic on/off functionality, because the cached [`call_model_generate_method`](https://github.com/normal-computing/outlines/blob/2af0b5ab05182ae86fdd931dd463b1c6dd986136/outlines/models/hf_transformers.py#L65) is constructed at module load time.  If we construct `call_model_generate_method` or apply the `cache` decorator to it [when completers `call` functions are constructed](https://github.com/normal-computing/outlines/blob/2af0b5ab05182ae86fdd931dd463b1c6dd986136/outlines/models/hf_transformers.py#L49), we might be able to remove that restriction.

File: outlines/__init__.py [hunk 3]
```diff
@@ -1,10 +1,38 @@
...
+    np.random.seed(seed)
+
+    with suppress(ImportError):
+        import torch
+
+        torch.manual_seed(seed)
+        torch.cuda.manual_seed_all(seed)
+        # ^^ safe to call this function even if cuda is not available
+
+    with suppress(ImportError):
```
[the remark this answers is not in the document]
Comment (brandonwillard) [reply outlines/__init__.py:35]: It was in the reference `set_seed` implementation linked above.

**Comment by rlouf** [comment]
I think the new `Sequence` solves this problem, no?

**Comment by brandonwillard** [comment]
> I think the new `Sequence` solves this problem, no?

Yeah, in large part it should.
